### PR TITLE
Restrict subscribing user to the same bank service more than once #44

### LINKED ISSUE #44: Restrict subscribing user to the same bank service more than once
Restrict subscribing users to the same bank service more than once if the service's periodicity is recurring.
Add the new check for subscription API for this business logic.
But if the subscription was previously canceled then allow a user to subscribe again.

### PR DESCRIPTION
**Pull request by andy-sam**
fixes #44

### CODE REVIEW

File: src/main/resources/db/changelog/changelog-2021-07-20-SubscriptionCanceling.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+databaseChangeLog:
+  - changeSet:
+      id: subscription-table-field-for-canceling-date
+      author: andy-sam
+      changes:
+        - addColumn:
+            columns:
+              - column:
+                  name: canceled_date
+                  type: date
```
Comment (alex-anakin) [inline src/main/resources/db/changelog/changelog-2021-07-20-SubscriptionCanceling.yaml:10]: date -> datetime

File: src/main/java/com/codesarge/banking/controllers/UserController.java [hunk 2]
```diff
@@ -59,4 +62,14 @@ public void deleteUser(@PathVariable("id") Long id) {
...
+
+    @GetMapping("users/{id}/subscriptions")
+    @ResponseBody
+    public List<SubscriptionDto> findAllSubscriptionsByUserId(
+            @PathVariable("id") Long id,
+            @RequestParam(required = false) Boolean active) {
+        log.info("Subscriptions list for user {} is requested", id);
+        return userService.findAllSubscriptionsByUserId(id, active);
+    }
+
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/controllers/UserController.java:74]: redundant empty line, please remove

File: src/main/java/com/codesarge/banking/entities/Subscription.java [hunk 3]
```diff
@@ -28,20 +29,25 @@
...
+    @Column
+    private LocalDateTime canceledDate;
+
     public Subscription(BankService bankService, Account account) {
         this.bankService = bankService;
         this.account = account;
         this.createdDate = LocalDateTime.now();
     }
+
+
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/entities/Subscription.java:52]: redundant empty lines, please remove

File: src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java [hunk 4]
```diff
@@ -8,4 +8,12 @@
 
 public interface SubscriptionRepository extends JpaRepository<Subscription, Long> {
     List<Subscription> findSubscriptionsByBankServicePeriodicity(Periodicity periodicity);
+
+    List<Subscription> findSubscriptionsByAccountId(Long accountId);
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java:12]: unused method, please remove

**Review by alex-anakin — CHANGES_REQUESTED** [review]

File: src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java [hunk 5]
```diff
@@ -8,4 +8,12 @@
...
+
+    List<Subscription> findSubscriptionsByAccountUserId(Long accountId);
```
[the remark this answers is not in the document]
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java:14]: done

File: src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java [hunk 6]
```diff
@@ -8,4 +8,12 @@
...
+
+    List<Subscription> findSubscriptionsByAccountUserIdAndCanceledDateIsNull(Long accountId);
```
[the remark this answers is not in the document]
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java:16]: done

File: src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java [hunk 7]
```diff
@@ -8,4 +8,12 @@
...
+
+    List<Subscription> findSubscriptionsByAccountUserIdAndCanceledDateIsNotNull(Long accountId);
```
[the remark this answers is not in the document]
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java:18]: done

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java [hunk 8]
```diff
@@ -38,6 +44,19 @@ public SubscriptionDto makeSubscription(SubscriptionRequestDto subscriptionReque
 //        checking if the account belongs to the correct user
         if (!account.getUser().getId().equals(subscriptionRequestDto.getUserId()))
             return null;
+//        checking if the user is subscribed to the same recurring bank service
+        List<Subscription> allSubscriptionsOfUser = subscriptionRepository
+                .findSubscriptionsByAccountUserId(account.getId());
```
[the remark this answers is not in the document]
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/service/SubscriptionService.java:49]: done

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java [hunk 9]
```diff
@@ -38,6 +44,19 @@ public SubscriptionDto makeSubscription(SubscriptionRequestDto subscriptionReque
...
+        for (Subscription s : allSubscriptionsOfUser) {
```
[the remark this answers is not in the document]
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/service/SubscriptionService.java:50]: done

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java [hunk 10]
```diff
@@ -38,6 +44,19 @@ public SubscriptionDto makeSubscription(SubscriptionRequestDto subscriptionReque
...
+            if (s.getBankService().getId().equals(subscriptionRequestDto.getBankServiceId())) {
+                if (s.getBankService().getPeriodicity().equals(Periodicity.WEEKLY) || s
```
[the remark this answers is not in the document]
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/service/SubscriptionService.java:52]: done

File: src/main/resources/db/changelog/changelog-2021-07-20-SubscriptionCanceling.yaml (hunk 1 above)
Comment (andy-sam) [reply src/main/resources/db/changelog/changelog-2021-07-20-SubscriptionCanceling.yaml:10]: done

File: src/main/java/com/codesarge/banking/controllers/UserController.java (hunk 2 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/controllers/UserController.java:74]: done

File: src/main/java/com/codesarge/banking/entities/Subscription.java (hunk 3 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/entities/Subscription.java:52]: done

File: src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java (hunk 4 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java:12]: done

**Review by alex-anakin — APPROVED** [review]

**Pull request merged**